File: ts/scripts/gencodes.mjs
#!/usr/bin/env node
// Generate the TypeScript strictspec error-code catalogue module.
//
// Parses .stricttools/docs/appendix-error-codes.md (the single normative source for the
// error-code catalogue) and emits src/codes.generated.ts. This mirrors the Go
// generator (go/tools/gencodes) and the Python generator
// (python/scripts/gencodes.py): the appendix is the only writer of the
// catalogue, and a freshness test regenerates and byte-compares (drift =
// failure).
//
// Usage:
//   node scripts/gencodes.mjs [--check]
//
// With no flags it auto-locates the repo root (the ancestor containing
// .stricttools/docs/appendix-error-codes.md) and derives both paths. --check regenerates in
// memory and exits non-zero if the on-disk file differs (the freshness gate).

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const SPEC_REL = ".stricttools/docs/appendix-error-codes.md";
const OUT_REL = "ts/src/codes.generated.ts";

const CODE_ROW_RE = /^\| `STRICTSPEC_/;
const AREA_ROW_RE = /^\| `([A-Z]+)` \|/;
const AREA_HEADING_RE = /^## 3\. /;
const NEXT_HEADING_RE = /^## /;
const PLACEHOLDER_RE = /\{(\w+)\}/g;
const SLOT_ENTRY_RE = /^([A-Za-z_]\w*):\s*(.+)$/;
const BACKTICK_CODE_RE = /^`(STRICTSPEC_[A-Z0-9_]+)`$/;

function findRepoRoot(start) {
	let d = start;
	for (;;) {
		if (existsSync(join(d, SPEC_REL))) {
			return d;
		}
		const parent = dirname(d);
		if (parent === d) {
			throw new Error(`gencodes: could not locate ${SPEC_REL} in any ancestor`);
		}
		d = parent;
	}
}

function splitCells(row) {
	const cells = [];
	let cur = "";
	let i = 0;
	const n = row.length;
	while (i < n) {
		const ch = row[i];
		if (ch === "\\" && i + 1 < n && row[i + 1] === "|") {
			cur += "|";
			i += 2;
			continue;
		}
		if (ch === "|") {
			cells.push(cur);
			cur = "";
			i += 1;
			continue;
		}
		cur += ch;
		i += 1;
	}
	cells.push(cur);
	let out = cells;
	if (out.length >= 2) {
		out = out.slice(1, -1);
	}
	return out.map((c) => c.trim());
}

function parseAreas(lines) {
	const areas = [];
	let inSection = false;
	for (const line of lines) {
		if (AREA_HEADING_RE.test(line)) {
			inSection = true;
			continue;
		}
		if (inSection && NEXT_HEADING_RE.test(line)) {
			break;
		}
		if (!inSection) {
			continue;
		}
		const m = AREA_ROW_RE.exec(line);
		if (m) {
			areas.push(m[1]);
		}
	}
	if (areas.length === 0) {
		throw new Error(
			"gencodes: could not parse the closed area set from section 3",
		);
	}
	return areas.sort();
}

function parseCodeCell(cell) {
	const m = BACKTICK_CODE_RE.exec(cell);
	if (m === null) {
		throw new Error(`gencodes: malformed code cell ${JSON.stringify(cell)}`);
	}
	return m[1];
}

function parseTemplateCell(cell) {
	if (cell.length < 2 || cell[0] !== "`" || cell[cell.length - 1] !== "`") {
		throw new Error(
			`gencodes: template cell not backtick-delimited: ${JSON.stringify(cell)}`,
		);
	}
	return cell.slice(1, -1);
}

function areaOf(code, areaSet) {
	const parts = code.split("_");
	if (parts.length < 3 || parts[0] !== "STRICTSPEC") {
		throw new Error(
			`gencodes: code ${JSON.stringify(code)} is not STRICTSPEC_<AREA>_<NAME>`,
		);
	}
	const area = parts[1];
	if (!areaSet.has(area)) {
		throw new Error(
			`gencodes: code ${JSON.stringify(code)} has area ${JSON.stringify(area)} outside the closed set`,
		);
	}
	return area;
}

const SCALAR_SLOT_TYPES = {
	string: "String",
	int: "Int",
	code: "Code",
	identifier: "Identifier",
	version: "Version",
	path: "Path",
	value: "Value",
};

function scalarSlotType(token) {
	const t = SCALAR_SLOT_TYPES[token];
	if (t === undefined) {
		throw new Error(`gencodes: unknown slot type ${JSON.stringify(token)}`);
	}
	return t;
}

function parseSlotType(name, typeTextIn) {
	const typeText = typeTextIn
		.replaceAll("\\<", "<")
		.replaceAll("\\>", ">")
		.trim();
	if (typeText.startsWith("list<") && typeText.endsWith(">")) {
		const elem = typeText.slice("list<".length, -1);
		return { name, typ: "List", elemType: scalarSlotType(elem) };
	}
	return { name, typ: scalarSlotType(typeText) };
}

function parseSlotsCell(cell) {
	const out = {};
	const c = cell.trim();
	if (c === "" || c === "—" || c === "-") {
		return out;
	}
	for (const partIn of c.split(",")) {
		const part = partIn.trim();
		if (part === "") {
			continue;
		}
		const m = SLOT_ENTRY_RE.exec(part);
		if (m === null) {
			throw new Error(
				`gencodes: malformed slot declaration ${JSON.stringify(part)}`,
			);
		}
		out[m[1]] = parseSlotType(m[1], m[2].trim());
	}
	return out;
}

function placeholderOrder(template) {
	const order = [];
	const seen = new Set();
	for (const m of template.matchAll(PLACEHOLDER_RE)) {
		const name = m[1];
		if (seen.has(name)) {
			continue;
		}
		seen.add(name);
		order.push(name);
	}
	return { order, seen };
}

function resolveSlots(code, template, declared) {
	const { order, seen } = placeholderOrder(template);
	for (const name of Object.keys(declared)) {
		if (!seen.has(name)) {
			throw new Error(
				`gencodes: code ${code} declares slot ${JSON.stringify(name)} the template does not reference`,
			);
		}
	}
	const slots = [];
	for (const name of order) {
		if (name in declared) {
			slots.push(declared[name]);
			continue;
		}
		if (name === "path") {
			slots.push({ name: "path", typ: "Path" });
		} else if (name === "suggestion") {
			slots.push({ name: "suggestion", typ: "String" });
		} else {
			throw new Error(
				`gencodes: code ${code}: placeholder {${name}} has no declared slot type`,
			);
		}
	}
	return slots;
}

function parseSpec(src) {
	const lines = src.split("\n");
	const areas = parseAreas(lines);
	const areaSet = new Set(areas);

	const entries = [];
	const seen = new Set();
	let lineno = 0;
	for (const line of lines) {
		lineno += 1;
		if (!CODE_ROW_RE.test(line)) {
			continue;
		}
		const cells = splitCells(line);
		if (cells.length < 3) {
			throw new Error(
				`gencodes: line ${lineno}: expected >=3 cells, got ${cells.length}`,
			);
		}
		const code = parseCodeCell(cells[0]);
		if (seen.has(code)) {
			throw new Error(`gencodes: line ${lineno}: duplicate code ${code}`);
		}
		seen.add(code);
		const template = parseTemplateCell(cells[1]);
		const area = areaOf(code, areaSet);
		const declared = parseSlotsCell(cells[2]);
		const slots = resolveSlots(code, template, declared);
		entries.push({ code, area, template, slots });
	}
	if (entries.length === 0) {
		throw new Error("gencodes: no code rows found");
	}
	entries.sort((a, b) => (a.code < b.code ? -1 : a.code > b.code ? 1 : 0));
	return { entries, areas };
}

function q(s) {
	return JSON.stringify(s);
}

function render(entries, areas) {
	const out = [];
	out.push(`/* eslint-disable */
// The generated strictspec error-code catalogue.
//
// Code generated by scripts/gencodes.mjs from ${SPEC_REL}. DO NOT EDIT.
//
// Every STRICTSPEC_* code with its area, message template, and declared slots,
// parsed from the single normative source. Hand-transcription is forbidden; a
// freshness test regenerates and byte-compares (drift = failure).

export enum SlotType {
\tString = 0,
\tInt = 1,
\tCode = 2,
\tIdentifier = 3,
\tVersion = 4,
\tPath = 5,
\tValue = 6,
\tList = 7,
}

export interface SlotSpec {
\treadonly name: string;
\treadonly type: SlotType;
\treadonly elemType?: SlotType;
}

export interface Entry {
\treadonly code: string;
\treadonly area: string;
\treadonly template: string;
\treadonly slots: readonly SlotSpec[];
}

export function lookup(code: string): Entry | undefined {
\treturn CATALOGUE.get(code);
}

export function allEntries(): Entry[] {
\treturn [...CATALOGUE.keys()].sort().map((k) => CATALOGUE.get(k) as Entry);
}

// The closed area set (appendix-error-codes.md section 3).
export const AREAS: readonly string[] = [
`);
	for (const a of areas) {
		out.push(`\t${q(a)},\n`);
	}
	out.push("];\n\nexport const CATALOGUE: Map<string, Entry> = new Map([\n");
	for (const e of entries) {
		out.push(`\t[\n\t\t${q(e.code)},\n\t\t{\n`);
		out.push(`\t\t\tcode: ${q(e.code)},\n`);
		out.push(`\t\t\tarea: ${q(e.area)},\n`);
		out.push(`\t\t\ttemplate: ${q(e.template)},\n`);
		if (e.slots.length === 0) {
			out.push("\t\t\tslots: [],\n");
		} else {
			out.push("\t\t\tslots: [\n");
			for (const s of e.slots) {
				if (s.elemType !== undefined) {
					out.push(
						`\t\t\t\t{ name: ${q(s.name)}, type: SlotType.${s.typ}, elemType: SlotType.${s.elemType} },\n`,
					);
				} else {
					out.push(
						`\t\t\t\t{ name: ${q(s.name)}, type: SlotType.${s.typ} },\n`,
					);
				}
			}
			out.push("\t\t\t],\n");
		}
		out.push("\t\t},\n\t],\n");
	}
	out.push("]);\n");
	return out.join("");
}

function generate(repoRoot) {
	const specPath = join(repoRoot, SPEC_REL);
	const src = readFileSync(specPath, "utf-8");
	const { entries, areas } = parseSpec(src);
	return { content: render(entries, areas), count: entries.length };
}

function main() {
	const check = process.argv.includes("--check");
	const root = findRepoRoot(dirname(fileURLToPath(import.meta.url)));
	const outPath = join(root, OUT_REL);
	const { content, count } = generate(root);

	if (check) {
		const existing = existsSync(outPath) ? readFileSync(outPath, "utf-8") : "";
		if (existing !== content) {
			process.stderr.write(
				`gencodes: ${outPath} is STALE; run scripts/gencodes.mjs\n`,
			);
			process.exit(1);
		}
		process.stdout.write(`gencodes: ${outPath} is fresh (${count} codes)\n`);
		return;
	}
	writeFileSync(outPath, content, "utf-8");
	process.stdout.write(`gencodes: wrote ${count} codes to ${outPath}\n`);
}

main();

File: python/scripts/gencodes.py
#!/usr/bin/env python3
"""Generate the Python strictspec error-code catalogue module.

Parses .stricttools/docs/appendix-error-codes.md (the single normative source for the
error-code catalogue) and emits src/strictspec/_codes.py. This mirrors the Go
generator go/tools/gencodes: the appendix is the only writer of the catalogue,
and a freshness test regenerates and byte-compares (drift = test failure).

Usage:
    python scripts/gencodes.py [--spec PATH] [--out PATH] [--check]

With no flags it auto-locates the repo root (the ancestor containing
.stricttools/docs/appendix-error-codes.md) and derives both paths. --check regenerates in
memory and exits non-zero if the on-disk file differs (the freshness gate).
"""

from __future__ import annotations

import argparse
import re
import sys
from dataclasses import dataclass
from pathlib import Path

SPEC_REL = ".stricttools/docs/appendix-error-codes.md"
OUT_REL = "python/src/strictspec/_codes.py"

_CODE_ROW_RE = re.compile(r"^\| `STRICTSPEC_")
_AREA_ROW_RE = re.compile(r"^\| `([A-Z]+)` \|")
_AREA_HEADING_RE = re.compile(r"^## 3\. ")
_NEXT_HEADING_RE = re.compile(r"^## ")
_PLACEHOLDER_RE = re.compile(r"\{(\w+)\}")
_SLOT_ENTRY_RE = re.compile(r"^([A-Za-z_]\w*):\s*(.+)$")
_BACKTICK_CODE_RE = re.compile(r"^`(STRICTSPEC_[A-Z0-9_]+)`$")


@dataclass
class SlotSpecData:
    name: str
    typ: str  # SlotType member name
    elem_type: str | None = None  # for list types


@dataclass
class EntryData:
    code: str
    area: str
    template: str
    slots: list[SlotSpecData]


def find_repo_root(start: Path) -> Path:
    d = start.resolve()
    while True:
        if (d / SPEC_REL).exists():
            return d
        if d.parent == d:
            raise SystemExit(f"gencodes: could not locate {SPEC_REL} in any ancestor")
        d = d.parent


def split_cells(row: str) -> list[str]:
    cells: list[str] = []
    cur: list[str] = []
    i = 0
    n = len(row)
    while i < n:
        ch = row[i]
        if ch == "\\" and i + 1 < n and row[i + 1] == "|":
            cur.append("|")
            i += 2
            continue
        if ch == "|":
            cells.append("".join(cur))
            cur = []
            i += 1
            continue
        cur.append(ch)
        i += 1
    cells.append("".join(cur))
    if len(cells) >= 2:
        cells = cells[1:-1]
    return [c.strip() for c in cells]


def parse_areas(lines: list[str]) -> list[str]:
    areas: list[str] = []
    in_section = False
    for line in lines:
        if _AREA_HEADING_RE.match(line):
            in_section = True
            continue
        if in_section and _NEXT_HEADING_RE.match(line):
            break
        if not in_section:
            continue
        m = _AREA_ROW_RE.match(line)
        if m:
            areas.append(m.group(1))
    if not areas:
        raise SystemExit("gencodes: could not parse the closed area set from section 3")
    return sorted(areas)


def parse_code_cell(cell: str) -> str:
    m = _BACKTICK_CODE_RE.match(cell)
    if m is None:
        raise SystemExit(f"gencodes: malformed code cell {cell!r}")
    return m.group(1)


def parse_template_cell(cell: str) -> str:
    if len(cell) < 2 or cell[0] != "`" or cell[-1] != "`":
        raise SystemExit(f"gencodes: template cell not backtick-delimited: {cell!r}")
    return cell[1:-1]


def area_of(code: str, area_set: set[str]) -> str:
    parts = code.split("_", 2)
    if len(parts) < 3 or parts[0] != "STRICTSPEC":
        raise SystemExit(f"gencodes: code {code!r} is not STRICTSPEC_<AREA>_<NAME>")
    area = parts[1]
    if area not in area_set:
        raise SystemExit(f"gencodes: code {code!r} has area {area!r} outside the closed set")
    return area


_SCALAR_SLOT_TYPES = {
    "string": "STRING",
    "int": "INT",
    "code": "CODE",
    "identifier": "IDENTIFIER",
    "version": "VERSION",
    "path": "PATH",
    "value": "VALUE",
}


def scalar_slot_type(token: str) -> str:
    if token not in _SCALAR_SLOT_TYPES:
        raise SystemExit(f"gencodes: unknown slot type {token!r}")
    return _SCALAR_SLOT_TYPES[token]


def parse_slot_type(name: str, type_text: str) -> SlotSpecData:
    type_text = type_text.replace(r"\<", "<").replace(r"\>", ">").strip()
    if type_text.startswith("list<") and type_text.endswith(">"):
        elem = type_text[len("list<") : -1]
        return SlotSpecData(name=name, typ="LIST", elem_type=scalar_slot_type(elem))
    return SlotSpecData(name=name, typ=scalar_slot_type(type_text))


def parse_slots_cell(cell: str) -> dict[str, SlotSpecData]:
    out: dict[str, SlotSpecData] = {}
    cell = cell.strip()
    if cell in ("", "—", "-"):
        return out
    for part in cell.split(","):
        part = part.strip()
        if part == "":
            continue
        m = _SLOT_ENTRY_RE.match(part)
        if m is None:
            raise SystemExit(f"gencodes: malformed slot declaration {part!r}")
        out[m.group(1)] = parse_slot_type(m.group(1), m.group(2).strip())
    return out


def resolve_slots(code: str, template: str, declared: dict[str, SlotSpecData]) -> list[SlotSpecData]:
    order: list[str] = []
    seen: set[str] = set()
    for m in _PLACEHOLDER_RE.finditer(template):
        name = m.group(1)
        if name in seen:
            continue
        seen.add(name)
        order.append(name)
    for name in declared:
        if name not in seen:
            raise SystemExit(
                f"gencodes: code {code} declares slot {name!r} the template does not reference"
            )
    slots: list[SlotSpecData] = []
    for name in order:
        if name in declared:
            slots.append(declared[name])
            continue
        if name == "path":
            slots.append(SlotSpecData(name="path", typ="PATH"))
        elif name == "suggestion":
            slots.append(SlotSpecData(name="suggestion", typ="STRING"))
        else:
            raise SystemExit(
                f"gencodes: code {code}: placeholder {{{name}}} has no declared slot type"
            )
    return slots


def parse(src: str) -> tuple[list[EntryData], list[str]]:
    lines = src.split("\n")
    areas = parse_areas(lines)
    area_set = set(areas)

    entries: list[EntryData] = []
    seen: set[str] = set()
    for lineno, line in enumerate(lines, 1):
        if not _CODE_ROW_RE.match(line):
            continue
        cells = split_cells(line)
        if len(cells) < 3:
            raise SystemExit(f"gencodes: line {lineno}: expected >=3 cells, got {len(cells)}")
        code = parse_code_cell(cells[0])
        if code in seen:
            raise SystemExit(f"gencodes: line {lineno}: duplicate code {code}")
        seen.add(code)
        template = parse_template_cell(cells[1])
        area = area_of(code, area_set)
        declared = parse_slots_cell(cells[2])
        slots = resolve_slots(code, template, declared)
        entries.append(EntryData(code=code, area=area, template=template, slots=slots))
    if not entries:
        raise SystemExit("gencodes: no code rows found")
    entries.sort(key=lambda e: e.code)
    return entries, areas


_HEADER = '''"""The generated strictspec error-code catalogue.

Code generated by scripts/gencodes.py from {spec}. DO NOT EDIT.

Every STRICTSPEC_* code with its area, message template, and declared slots,
parsed from the single normative source. Hand-transcription is forbidden; a
freshness test regenerates and byte-compares (drift = failure).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum


class SlotType(IntEnum):
    """One of the slot-type vocabulary of appendix-error-codes.md sec.2."""

    STRING = 0
    INT = 1
    CODE = 2
    IDENTIFIER = 3
    VERSION = 4
    PATH = 5
    VALUE = 6
    LIST = 7

    def __str__(self) -> str:
        return _SLOT_TYPE_NAMES.get(self, "unknown")


_SLOT_TYPE_NAMES = {{
    SlotType.STRING: "string",
    SlotType.INT: "int",
    SlotType.CODE: "code",
    SlotType.IDENTIFIER: "identifier",
    SlotType.VERSION: "version",
    SlotType.PATH: "path",
    SlotType.VALUE: "value",
    SlotType.LIST: "list",
}}


@dataclass(frozen=True)
class SlotSpec:
    """One declared slot: its placeholder name and declared type. For LIST,
    elem_type is the element type (e.g. list<string>).
    """

    name: str
    type: SlotType
    elem_type: SlotType | None = None


@dataclass(frozen=True)
class Entry:
    """One catalogue row: the code, its area, the pinned message template, and
    the slots the template interpolates in placeholder order.
    """

    code: str
    area: str
    template: str
    slots: tuple[SlotSpec, ...] = ()


def lookup(code: str) -> Entry | None:
    """Return the catalogue entry for a code, or None."""
    return CATALOGUE.get(code)


def all_entries() -> list[Entry]:
    """Every catalogue entry, sorted by code."""
    return [CATALOGUE[k] for k in sorted(CATALOGUE)]
'''


def render(entries: list[EntryData], areas: list[str], spec_rel: str) -> str:
    out: list[str] = []
    out.append(_HEADER.format(spec=spec_rel))
    out.append("\n\n")
    out.append("# The closed area set (appendix-error-codes.md section 3).\n")
    out.append("AREAS = (\n")
    for a in areas:
        out.append(f"    {a!r},\n")
    out.append(")\n\n")
    out.append("CATALOGUE: dict[str, Entry] = {\n")
    for e in entries:
        out.append(f"    {e.code!r}: Entry(\n")
        out.append(f"        code={e.code!r},\n")
        out.append(f"        area={e.area!r},\n")
        out.append(f"        template={e.template!r},\n")
        if not e.slots:
            out.append("        slots=(),\n")
        else:
            out.append("        slots=(\n")
            for s in e.slots:
                if s.elem_type is not None:
                    out.append(
                        f"            SlotSpec({s.name!r}, SlotType.{s.typ}, SlotType.{s.elem_type}),\n"
                    )
                else:
                    out.append(f"            SlotSpec({s.name!r}, SlotType.{s.typ}),\n")
            out.append("        ),\n")
        out.append("    ),\n")
    out.append("}\n")
    return "".join(out)


def generate(repo_root: Path) -> str:
    spec_path = repo_root / SPEC_REL
    src = spec_path.read_text(encoding="utf-8")
    entries, areas = parse(src)
    return render(entries, areas, SPEC_REL)


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--spec", default="")
    ap.add_argument("--out", default="")
    ap.add_argument("--check", action="store_true", help="fail if the on-disk file is stale")
    args = ap.parse_args()

    root = find_repo_root(Path(__file__).parent)
    spec_path = Path(args.spec) if args.spec else root / SPEC_REL
    out_path = Path(args.out) if args.out else root / OUT_REL

    src = spec_path.read_text(encoding="utf-8")
    entries, areas = parse(src)
    content = render(entries, areas, SPEC_REL)

    if args.check:
        existing = out_path.read_text(encoding="utf-8") if out_path.exists() else ""
        if existing != content:
            print(f"gencodes: {out_path} is STALE; run scripts/gencodes.py", file=sys.stderr)
            return 1
        print(f"gencodes: {out_path} is fresh ({len(entries)} codes)")
        return 0

    out_path.write_text(content, encoding="utf-8")
    print(f"gencodes: wrote {len(entries)} codes to {out_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
